sycl/source/CMakeLists.txt

```diff
@@ -4,101 +4,148 @@
+    "detail/usm/clusm.cpp"
+    "detail/usm/usm_dispatch.cpp"
+    "detail/usm/usm_impl.cpp"
+    "detail/util.cpp"
+    "context.cpp"
+    "device.cpp"
+    "device_selector.cpp"
+    "event.cpp"
+    "exception.cpp"
+    "exception_list.cpp"
+    "half_type.cpp"
+    "kernel.cpp"
+    "platform.cpp"
+    "queue.cpp"
+    "ordered_queue.cpp"
+    "sampler.cpp"
+    "stream.cpp"
+    "spirv_ops.cpp"
 )
 
-set_target_properties(sycl PROPERTIES LINKER_LANGUAGE CXX)
+add_sycl_rt_library(sycl ${SYCL_SOURCES})
 
 if (MSVC)
-    target_compile_definitions(sycl PRIVATE __SYCL_BUILD_SYCL_DLL )
-endif()
-target_include_directories(sycl PRIVATE "${sycl_inc_dir}")
-target_link_libraries(sycl
-    PRIVATE OpenCL::Headers
-    PRIVATE ${OpenCL_LIBRARIES}
-)
-if (SYCL_USE_LIBCXX)
-    if ((CMAKE_CXX_COMPILER_ID STREQUAL "GNU") OR
-        (CMAKE_CXX_COMPILER_ID STREQUAL "Clang"))
-        target_compile_options(sycl PRIVATE -nostdinc++)
-        if ((NOT (DEFINED SYCL_LIBCXX_INCLUDE_PATH)) OR (NOT (DEFINED SYCL_LIBCXX_LIBRARY_PATH)))
-            message(FATAL_ERROR "When building with libc++ SYCL_LIBCXX_INCLUDE_PATHS and"
-                                "SYCL_LIBCXX_LIBRARY_PATH should be set")
-        endif()
-        target_include_directories(sycl PRIVATE "${SYCL_LIBCXX_INCLUDE_PATH}")
-        target_link_libraries(sycl PRIVATE "-L${SYCL_LIBCXX_LIBRARY_PATH}" -nodefaultlibs -lc++ -lc++abi -lm -lc -lgcc_s -lgcc)
-    else()
-        message(FATAL_ERROR "Build with libc++ is not yet supported for this compiler")
+  # MSVC provides two incompatible build variants for its CRT: release and debug
+  # To avoid potential issues in user code we also need to provide two kinds
+  # of SYCL Runtime Library for release and debug configurations.
+  set(SYCL_CXX_FLAGS "")
+  if (CMAKE_BUILD_TYPE MATCHES "Debug")
+    set(SYCL_CXX_FLAGS "${CMAKE_CXX_FLAGS_DEBUG}")
+    string(REPLACE "/MDd" "" SYCL_CXX_FLAGS "${SYCL_CXX_FLAGS}")
+    string(REPLACE "/MTd" "" SYCL_CXX_FLAGS "${SYCL_CXX_FLAGS}")
+  else()
+    if (CMAKE_BUILD_TYPE MATCHES "Release")
+      set(SYCL_CXX_FLAGS "${CMAKE_CXX_FLAGS_RELEASE}")
+    elseif (CMAKE_BUILD_TYPE MATCHES "RelWithDebInfo")
+      set(SYCL_CXX_FLAGS "${CMAKE_CXX_FLAGS_MINSIZEREL}")
+    elseif (CMAKE_BUILD_TYPE MATCHES "MinSizeRel")
+      set(SYCL_CXX_FLAGS "${CMAKE_CXX_FLAGS_RELWITHDEBINFO}")
     endif()
-else()
+    string(REPLACE "/MD" "" SYCL_CXX_FLAGS "${SYCL_CXX_FLAGS}")
+    string(REPLACE "/MT" "" SYCL_CXX_FLAGS "${SYCL_CXX_FLAGS}")
+  endif()
 
-# Workaround for bug in GCC version 5 and higher.
-# More information https://bugs.launchpad.net/ubuntu/+source/gcc-5/+bug/1568899
-if (CMAKE_CXX_COMPILER_ID STREQUAL "GNU" AND
-    CMAKE_CXX_COMPILER_VERSION VERSION_GREATER 5.0)
-  target_link_libraries(sycl PRIVATE gcc_s gcc)
-endif()
+  # target_compile_options requires list of options, not a string
+  string(REPLACE " " ";" SYCL_CXX_FLAGS "${SYCL_CXX_FLAGS}")
+
+  set(SYCL_CXX_FLAGS_RELEASE "${SYCL_CXX_FLAGS};/MD")
```

Comment:
Can a DLL built with /MD flag be used by a program built with /MT?

Reply:
Yes, it should work fine. 
Actually, I think providing a library variant built with /MT (if that's what your're implying) may cause a lot more trouble. Let's say, we link SYCL RT DLL against CRT of version X. And user downloads pre-built compiler and runtime, develops an application and links it against SYCL RT and CRT of version Y. There are now 2 CRTs in app's address space. And they may be incompatible. Thus, I think providing only /MD version should be sufficient.

Reply:
> I think providing a library variant built with /MT (if that's what your're implying)

I did not suggest that. The only question is whether /MT for a user program works with sycl.dll or not. If it works fine, then please add another RUN line with /MT to the tests. If it doesn't - then we should explicitly state this requirement or provide a solution for /MT.